lib/dal/src/component.rs

```diff
@@ -3952,13 +3972,156 @@ impl Component {
+        for component_id in component_ids {
+            let maybe_component =
+                Component::try_get_by_id(&base_change_set_ctx, component_id).await?;
+            match maybe_component {
+                Some(_) => components.insert(component_id, true),
+                _ => components.insert(component_id, false),
+            };
+        }
+        Ok(components)
+    }
+
+    pub async fn into_frontend_type(
+        &self,
+        ctx: &DalContext,
+        change_status: ChangeStatus,
+    ) -> ComponentResult<SummaryDiagramComponent> {
+        let mut diagram_sockets: HashMap<SchemaVariantId, serde_json::Value> = HashMap::new();
+        let schema_variant = self.schema_variant(ctx).await?;
+
+        let sockets = match diagram_sockets.entry(schema_variant.id()) {
+            hash_map::Entry::Vacant(entry) => {
+                let (output_sockets, input_sockets) =
+                    SchemaVariant::list_all_sockets(ctx, schema_variant.id()).await?;
+
+                let mut sockets = vec![];
+
+                for socket in input_sockets {
+                    sockets.push(DiagramSocket {
+                        id: socket.id().to_string(),
+                        label: socket.name().to_string(),
+                        connection_annotations: socket.connection_annotations(),
+                        direction: DiagramSocketDirection::Input,
+                        max_connections: match socket.arity() {
+                            SocketArity::Many => None,
+                            SocketArity::One => Some(1),
+                        },
+                        is_required: Some(false),
+                        node_side: DiagramSocketNodeSide::Left,
+                    });
+                }
+
+                for socket in output_sockets {
+                    sockets.push(DiagramSocket {
+                        id: socket.id().to_string(),
+                        label: socket.name().to_string(),
+                        connection_annotations: socket.connection_annotations(),
+                        direction: DiagramSocketDirection::Output,
+                        max_connections: match socket.arity() {
+                            SocketArity::Many => None,
+                            SocketArity::One => Some(1),
+                        },
+                        is_required: Some(false),
+                        node_side: DiagramSocketNodeSide::Right,
+                    });
+                }
+
+                let socket_value = serde_json::to_value(sockets)?;
+
+                entry.insert(socket_value.to_owned());
+
+                socket_value
+            }
+            hash_map::Entry::Occupied(entry) => entry.get().to_owned(),
+        };
+        let schema = SchemaVariant::schema_for_schema_variant_id(ctx, schema_variant.id()).await?;
+
+        let default_schema_variant =
+            SchemaVariant::get_default_for_schema(ctx, schema.id()).await?;
```

Comment:
Are we always wanting to walk back to the default even though we already have a variant?

Reply:
We seem to use it for `can_be_upgraded`... I copy pasta'd this code.. just moved it from one place to another.